core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java

```diff
@@ -163,33 +178,128 @@ public boolean deleteDirectory(String path, DeleteOptions options) throws IOExce
                 + "recursive as true in order to delete non empty directories.", path);
         return false;
       }
-    } else {
-      // Delete children
-      UnderFileStatus[] pathsToDelete =
-          listInternal(path, ListOptions.defaults().setRecursive(true));
-      if (pathsToDelete == null) {
-        LOG.error("Unable to delete {} because listInternal returns null", path);
-        return false;
+      // Delete the directory itself
+      return deleteObject(stripPrefixIfPresent(convertToFolderName(path)));
+    }
+
+    // Delete children
+    DeleteBuffer deleteBuffer = new DeleteBuffer();
+    UnderFileStatus[] pathsToDelete = listInternal(path, ListOptions.defaults().setRecursive(true));
+    if (pathsToDelete == null) {
+      LOG.warn("Unable to delete {} because listInternal returns null", path);
+      return false;
+    }
+    for (UnderFileStatus pathToDelete : pathsToDelete) {
+      String pathKey = stripPrefixIfPresent(PathUtils.concatPath(path, pathToDelete.getName()));
+      if (pathToDelete.isDirectory()) {
+        deleteBuffer.add(convertToFolderName(pathKey));
+      } else {
+        deleteBuffer.add(pathKey);
       }
-      for (UnderFileStatus pathToDelete : pathsToDelete) {
-        // If we fail to deleteObject one file, stop
-        String pathKey = stripPrefixIfPresent(PathUtils.concatPath(path, pathToDelete.getName()));
-        if (pathToDelete.isDirectory()) {
-          if (!deleteObject(convertToFolderName(pathKey))) {
-            // If path is a directory, it is possible that it was not created through Alluxio and no
-            // zero-byte breadcrumb exists
-            LOG.warn("Failed to delete directory {}", pathToDelete.getName());
-          }
-        } else {
-          if (!deleteObject(pathKey)) {
-            LOG.error("Failed to delete file {}", pathToDelete.getName());
-            return false;
-          }
+    }
+    deleteBuffer.add(stripPrefixIfPresent(convertToFolderName(path)));
+    return deleteBuffer.getResult().size() == deleteBuffer.mEntriesAdded;
```

Comment:
correct, it relies on the underlying client's timeout and retry policy. there is not timeout / retry logic at the ufs level